### PR TITLE
Back to unmodified state with undo

### PR DESCRIPTION
**Pull request by tyfkda**
Text buffer keeps modified state even if a user undo all modifications,
because undo doesn't restore unmodified state.

This change counts how many times a buffer is modified,
and restore unmodified state if changes are undone.

### CODE REVIEW

File: src/history.rs [hunk 1]
```diff
@@ -53,23 +54,25 @@ impl History {
         })
     }
 
-    pub fn undo(&mut self, rows: &mut Vec<Row>) -> Option<(usize, usize, usize)> {
-        self.finish_ongoing_edit();
+    pub fn undo(&mut self, rows: &mut Vec<Row>) -> Option<(usize, usize, usize, bool)> {
+        let ongoing = self.finish_ongoing_edit();
```
Comment (rhysd) [inline src/history.rs:58]: I feel this variable name `ongoing` is not suitable. It represents wether some diff was applied or not. So I feel `edited` or `modified` is more suitable.

File: src/text_buffer.rs [hunk 2]
```diff
@@ -486,6 +495,8 @@ impl TextBuffer {
     }
 
     pub fn save(&mut self) -> std::result::Result<String, String> {
+        self.insert_undo_point();
```
Comment (rhysd) [inline src/text_buffer.rs:498]: Why is this necessary? Text is not modified on saving file.

And by this line, saving files multiple times without modifying text inserts multiple undo points.

File: src/text_buffer.rs [hunk 3]
```diff
@@ -130,7 +135,8 @@ impl TextBuffer {
             cy: 0,
             file,
             row,
-            modified: false,
+            modified_count: 0,
+            tmp_modified: false,
```
Comment (rhysd) [inline src/text_buffer.rs:139]: What is the reason to remain `modified` as `tmp_modified`? I thought replacing `modified` with `modified_count` is simpler

File: src/text_buffer.rs [hunk 4]
```diff
@@ -532,11 +544,23 @@ impl TextBuffer {
 
     pub fn undo(&mut self) -> bool {
         let state = self.history.undo(&mut self.row);
+        if let Some((_, _, _, ongoing)) = state {
+            if !ongoing {
```
Comment (rhysd) [inline src/text_buffer.rs:548]: Isn't this condition opposite? The bool value represents some diff appears or not derived from `finish_ongoing_edit()`. It means that modification occurred or not. So `true` means text was modified and `self_modified_count` should be updated.

File: src/text_buffer.rs [hunk 5]
```diff
@@ -532,11 +544,23 @@ impl TextBuffer {
...
+            }
+            self.tmp_modified = false;
+        }
         self.after_undoredo(state)
     }
 
     pub fn redo(&mut self) -> bool {
         let state = self.history.redo(&mut self.row);
+        if let Some((_, _, _, ongoing)) = state {
+            if !ongoing {
```
Comment (rhysd) [inline src/text_buffer.rs:559]: ditto

File: src/text_buffer.rs (hunk 3 above)
Comment (tyfkda) [reply src/text_buffer.rs:139]: `tmp_modified` is not same meaning as `modified`.
`tmp_modified` indicates that there are some changes not put into history.
In the case, `modified_count` might be `0`, but
I want to display `(modified)` in status bar.

File: src/text_buffer.rs (hunk 4 above)
Comment (tyfkda) [reply src/text_buffer.rs:548]: If there were some changes not put into history (`edited = true`),
these changes are canceled by undo,
so modified count keeps as is.

In the other hand, if `edited = false`,
then undo reverts last history
so modified count should be decremented.

File: src/text_buffer.rs (hunk 5 above)
Comment (tyfkda) [reply src/text_buffer.rs:559]: If there were some changes not put into history (edited = true),
these changes are put into history
and history after that are cleared.
So no redo occurred and modified count keeps as is.

On the other hand,
modified count should be incremented.

File: src/text_buffer.rs (hunk 2 above)
Comment (tyfkda) [reply src/text_buffer.rs:498]: It happens that `save` function is called but some ongoing changes exist.
I want to put these changes into history at the time,
so the text buffer is unmodified in saved state.

It seems no undo point inserted calling `insert_undo_point` function multiple times
because `self.inserted_undo` is `false`.

File: src/history.rs (hunk 1 above)
Comment (tyfkda) [reply src/history.rs]: Renamed to `edited`.

**Comment by tyfkda** [comment]
@rhysd Thank you for reviewing.
I renamed local variable, and replied to your comments.
Would you please take a look?

File: src/text_buffer.rs (hunk 2 above)
Comment (rhysd) [reply src/text_buffer.rs:498]: I got the point. Thanks for explanation. However, inserting undo point here means that an undo point is created even if saving file is cancelled. Can we move this line just before returning from this function?

Comment (rhysd) [reply src/text_buffer.rs:498]: OK, I got the point. Thanks for the explanation. I tried this change on my local and it worked fine.

File: src/text_buffer.rs (hunk 3 above)
Comment (rhysd) [reply src/text_buffer.rs:139]: I see. So I guess 'temporary' is similar to 'ongoing' here.

File: src/text_buffer.rs (hunk 4 above)
Comment (rhysd) [reply src/text_buffer.rs]: oh, ok, I got the point. I misunderstood the logic.
